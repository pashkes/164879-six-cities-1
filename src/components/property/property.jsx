import React, {PureComponent} from "react";
import PropType from "prop-types";
import {connect} from "react-redux";
import compose from 'recompose/compose';

import {
  getNearbyOffers,
  getCurrentCity,
  getCurrentOffer,
} from "./../../reducer/data/selectors";
import withLoadData from "./../../hocs/with-load-data/with-load-data";
import {Operation as DataOperation, ActionCreators} from "../../reducer/data/data";
import {toPercentRating} from "./../../utils";
import {getAuthorizationStatus} from "../../reducer/user/selectors";

import Layout from "./../layout/layout.jsx";
import Gallery from "./../gallery/gallery.jsx";
import Goods from "./../goods/goods.jsx";
import Reviews from "./../reviews/reviews.jsx";
import Map from "./../map/map.jsx";
import Offers from "./../offers/offers.jsx";
import ReviewForm from "./../review-form/review-form.jsx";
import FavoriteButton from "./../favorite-button/favorite-button.jsx";

export class Property extends PureComponent {
  constructor(props) {
    super(props);
  }

  componentDidMount() {
    const {setCurrentCity, currentOffer} = this.props;
    setCurrentCity(currentOffer.city.name);
  }

  render() {
    const {
      id,
      currentCity,
      offersOnMap,
      nearbyOffers,
      isAuthorizationRequired,
      currentOffer
    } = this.props;
    const {
      images,
      isPremium,
      title,
      isFavorite,
      rating,
      price,
      host: {
        avatarURL,
        isPro,
        name,
      },
      goods,
      description,
      maxAdults,
      bedrooms,
      type,
      location: {latitude, longitude},
    } = currentOffer;

    return (
      <Layout pageClasses={[`page`]}>
        <main className="page__main page__main--property">
          <section className="property">
            <div className="property__gallery-container container">
              <div className="property__gallery">
                <Gallery photos={images}/>
              </div>
            </div>
            <div className="property__container container">
              <div className="property__wrapper">
                {
                  isPremium &&
                  <div className="property__mark">
                    <span>Premium</span>
                  </div>
                }
                <div className="property__name-wrapper">
                  <h1 className="property__name">{title}</h1>
                  <FavoriteButton
                    id={id}
                    isFavorite={isFavorite}
                    prefixClass={`property`}
                    width={`31`}
                    height={`33`}
                  />
                </div>
                <div className="property__rating rating">
                  <div className="property__stars rating__stars">
                    <span style={{width: `${rating}%`}}/>
                    <span className="visually-hidden">Rating</span>
                  </div>
                  <span className="property__rating-value rating__value">{toPercentRating(rating)}</span>
                </div>
                <ul className="property__features">
                  <li className="property__feature property__feature--entire">{type}</li>
                  <li className="property__feature property__feature--bedrooms">{bedrooms} Bedrooms</li>
                  <li className="property__feature property__feature--adults">{maxAdults}</li>
                </ul>
                <div className="property__price">
                  <b className="property__price-value">&euro;{price}</b>
                  <span className="property__price-text">&nbsp;night</span>
                </div>
                <div className="property__inside">
                  <h2 className="property__inside-title">What&apos;s inside</h2>
                  <Goods list={goods}/>
                </div>
                <div className="property__host">
                  <h2 className="property__host-title">Meet the host</h2>
                  <div className="property__host-user user">
                    <div className="property__avatar-wrapper property__avatar-wrapper--pro user__avatar-wrapper">
                      <img className="property__avatar user__avatar" src={avatarURL} width="74" height="74" alt={name}/>
                    </div>
                    <span className="property__user-name">{name}</span>
                    {isPro && <span className="property__user-status">Pro</span>}
                  </div>
                  <div className="property__description">
                    <p className="property__text">{description}</p>
                  </div>
                </div>
                <section className="property__reviews reviews">
                  <Reviews id={id}/>
                  {!isAuthorizationRequired && <ReviewForm idCurrentOffer={id} />}
                </section>
              </div>
            </div>
            <section className="property__map map">
              <Map
                selectedOffer={[latitude, longitude]}
                currentCity={currentCity}
                coordinates={offersOnMap}
              />
            </section>
          </section>
          <div className="container">
            <section className="near-places places">
              <h2 className="near-places__title">Other places in the neighbourhood</h2>
              <Offers
                offers={nearbyOffers}
                classModOffers={[`near-places__list`]}
                classModCard={`near-places__card`}
              />
            </section>
          </div>
        </main>
      </Layout>
    );
  }
}

Property.propTypes = {
  id: PropType.number.isRequired,
  currentOffer: PropType.shape({
    images: PropType.array.isRequired,
    isPremium: PropType.bool.isRequired,
    title: PropType.string.isRequired,
    isFavorite: PropType.bool.isRequired,
    rating: PropType.number.isRequired,
    goods: PropType.array.isRequired,
    price: PropType.number.isRequired,
    city: PropType.shape({
      name: PropType.string.isRequired,
    }).isRequired,
    host: PropType.shape({
      avatarURL: PropType.string.isRequired,
      name: PropType.string.isRequired,
      isPro: PropType.bool.isRequired,
    }).isRequired,
    description: PropType.string.isRequired,
    maxAdults: PropType.number.isRequired,
    bedrooms: PropType.number.isRequired,
    type: PropType.string.isRequired,
    location: PropType.shape({
      latitude: PropType.number.isRequired,
      longitude: PropType.number.isRequired,
    }).isRequired
  }).isRequired,
  nearbyOffers: PropType.array.isRequired,
  currentCity: PropType.string.isRequired,
  offersOnMap: PropType.array.isRequired,
  isAuthorizationRequired: PropType.bool.isRequired,
  setCurrentCity: PropType.func.isRequired,
  currentOfferCoordinates: PropType.array,
};


const mapStateToProps = (state, ownProps) => {
  const nearbyOffers = getNearbyOffers(state, ownProps.id);

  return {
    isLoading: Boolean(nearbyOffers),
    currentOffer: getCurrentOffer(state, ownProps.id),
    nearbyOffers: nearbyOffers || [],
    offersOnMap: nearbyOffers.map((it) => [it.location.latitude, it.location.longitude]),
    currentCity: getCurrentCity(state),
    isAuthorizationRequired: getAuthorizationStatus(state),
  };
};

const mapDispatchToProps = (dispatch) => ({
  loadData: () => dispatch(DataOperation.loadOffers()),
  setCurrentCity: (city) => dispatch(ActionCreators.setCity(city)),
});


const property = compose(
    connect(mapStateToProps, mapDispatchToProps),
    withLoadData
);

export default property(Property);
