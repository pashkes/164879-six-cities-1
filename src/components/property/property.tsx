import * as React from "react";
import {connect} from "react-redux";
import compose from 'recompose/compose';

import {
  getNearbyOffers,
  getCurrentCity,
  getCurrentOffer,
} from "./../../reducer/data/selectors";
import withLoadData from "./../../hocs/with-load-data/with-load-data";
import {Operation as DataOperation, ActionCreators} from "../../reducer/data/data";
import {toPercentRating} from "./../../utils";
import {getAuthorizationStatus} from "../../reducer/user/selectors";

import Layout from "./../layout/layout";
import Gallery from "./../gallery/gallery";
import Goods from "./../goods/goods";
import Reviews from "./../reviews/reviews";
import Map from "./../map/map";
import Offers from "./../offers/offers";
import ReviewForm from "./../review-form/review-form";
import FavoriteButton from "./../favorite-button/favorite-button";
import {Offer} from "../../types";

interface Props {
  id: number,
  currentOffer: Offer,
  nearbyOffers: [number[]],
  currentCity: string,
  offersOnMap: [[number, number]],
  isAuthRequire: boolean,
  onSetCurrentCity: (city: string) => void,
  currentOfferCoordinates: number[],
}

export class Property extends React.PureComponent<Props> {
  componentDidMount() {
    const {onSetCurrentCity, currentOffer} = this.props;

    onSetCurrentCity(currentOffer.city.name);
  }

  render() {
    const {
      id,
      currentCity,
      offersOnMap,
      nearbyOffers,
      isAuthRequire,
      currentOffer
    } = this.props;

    const {
      images,
      isPremium,
      title,
      isFavorite,
      rating,
      price,
      host: {
        avatarURL,
        isPro,
        name,
      },
      goods,
      description,
      maxAdults,
      bedrooms,
      type,
      location: {latitude, longitude},
    } = currentOffer;

    return (
      <Layout title={title} pageClasses={`page`}>
        <main className="page__main page__main--property">
          <section className="property">
            <div className="property__gallery-container container">
              <div className="property__gallery">
                <Gallery photos={images}/>
              </div>
            </div>
            <div className="property__container container">
              <div className="property__wrapper">
                {
                  isPremium &&
                  <div className="property__mark">
                    <span>Premium</span>
                  </div>
                }
                <div className="property__name-wrapper">
                  <h1 className="property__name">{title}</h1>
                  <FavoriteButton
                    id={id}
                    isFavorite={isFavorite}
                    prefixClass={`property`}
                    width={`31`}
                    height={`33`}
                  />
                </div>
                <div className="property__rating rating">
                  <div className="property__stars rating__stars">
                    <span style={{width: `${rating}%`}}/>
                    <span className="visually-hidden">Rating</span>
                  </div>
                  <span className="property__rating-value rating__value">{toPercentRating(rating)}</span>
                </div>
                <ul className="property__features">
                  <li className="property__feature property__feature--entire">{type}</li>
                  <li className="property__feature property__feature--bedrooms">{bedrooms} Bedrooms</li>
                  <li className="property__feature property__feature--adults">{maxAdults}</li>
                </ul>
                <div className="property__price">
                  <b className="property__price-value">&euro;{price}</b>
                  <span className="property__price-text">&nbsp;night</span>
                </div>
                <div className="property__inside">
                  <h2 className="property__inside-title">What&apos;s inside</h2>
                  <Goods list={goods}/>
                </div>
                <div className="property__host">
                  <h2 className="property__host-title">Meet the host</h2>
                  <div className="property__host-user user">
                    <div className="property__avatar-wrapper property__avatar-wrapper--pro user__avatar-wrapper">
                      <img className="property__avatar user__avatar" src={avatarURL} width="74" height="74" alt={name}/>
                    </div>
                    <span className="property__user-name">{name}</span>
                    {isPro && <span className="property__user-status">Pro</span>}
                  </div>
                  <div className="property__description">
                    <p className="property__text">{description}</p>
                  </div>
                </div>
                <section className="property__reviews reviews">
                  <Reviews id={id}/>
                  {!isAuthRequire && <ReviewForm idCurrentOffer={id}/>}
                </section>
              </div>
            </div>
            <section className="property__map map">
              <Map
                selectedOffer={[latitude, longitude]}
                currentCity={currentCity}
                coordinates={offersOnMap}
              />
            </section>
          </section>
          <div className="container">
            <section className="near-places places">
              <h2 className="near-places__title">Other places in the neighbourhood</h2>
              <Offers
                offers={nearbyOffers}
                classModOffers={[`near-places__list`]}
                classModCard={`near-places__card`}
              />
            </section>
          </div>
        </main>
      </Layout>
    );
  }
}

const mapStateToProps = (state, ownProps) => {
  const nearbyOffers = getNearbyOffers(state, ownProps.id);

  return {
    isLoading: Boolean(nearbyOffers),
    currentOffer: getCurrentOffer(state, ownProps.id),
    nearbyOffers: nearbyOffers || [],
    offersOnMap: nearbyOffers.map((it) => [it.location.latitude, it.location.longitude]),
    currentCity: getCurrentCity(state),
    isAuthRequire: getAuthorizationStatus(state),
  };
};

const mapDispatchToProps = (dispatch) => ({
  onLoadData: () => dispatch(DataOperation.loadOffers()),
  onSetCurrentCity: (city) => dispatch(ActionCreators.setCity(city)),
});


const enhance = compose(
  connect(mapStateToProps, mapDispatchToProps),
  withLoadData
);

export default enhance(Property);
