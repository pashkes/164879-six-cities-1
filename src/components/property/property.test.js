import React from "react";
import {shallow, configure} from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import {Property} from './property.jsx';
configure({adapter: new Adapter()});


const propertyProps = {
  images: [``],
  isPremium: false,
  title: ``,
  isFavorite: false,
  rating: 0,
  goods: [``],
  price: 0,
  host: {
    avatarURL: ``,
    name: ``,
    isPro: false,
  },
  description: ``,
  maxAdults: 1,
  bedrooms: 1,
  type: ``,
};

it(`Property page renders correctly`, () => {
  const propertyPage = shallow(<Property
    id={`0`}
    targetOffer={propertyProps}
  />);
  expect(propertyPage).toMatchSnapshot();
});
