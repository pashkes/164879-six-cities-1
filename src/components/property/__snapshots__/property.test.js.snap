// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Property page renders correctly 1`] = `
<Layout
  pageClasses={
    Array [
      "page",
    ]
  }
>
  <main
    className="page__main page__main--property"
  >
    <section
      className="property"
    >
      <div
        className="property__gallery-container container"
      >
        <div
          className="property__gallery"
        >
          <Gallery
            photos={
              Array [
                "",
              ]
            }
          />
        </div>
      </div>
      <div
        className="property__container container"
      >
        <div
          className="property__wrapper"
        >
          <div
            className="property__name-wrapper"
          >
            <h1
              className="property__name"
            />
            <button
              className="property__bookmark-button button"
              type="button"
            >
              <svg
                className="property__bookmark-icon"
                height="33"
                width="31"
              >
                <use
                  xlinkHref="#icon-bookmark"
                />
              </svg>
              <span
                className="visually-hidden"
              >
                To bookmarks
              </span>
            </button>
          </div>
          <div
            className="property__rating rating"
          >
            <div
              className="property__stars rating__stars"
            >
              <span
                style={
                  Object {
                    "width": "0%",
                  }
                }
              />
              <span
                className="visually-hidden"
              >
                Rating
              </span>
            </div>
            <span
              className="property__rating-value rating__value"
            >
              0.0
            </span>
          </div>
          <ul
            className="property__features"
          >
            <li
              className="property__feature property__feature--entire"
            />
            <li
              className="property__feature property__feature--bedrooms"
            >
              1
               Bedrooms
            </li>
            <li
              className="property__feature property__feature--adults"
            >
              1
            </li>
          </ul>
          <div
            className="property__price"
          >
            <b
              className="property__price-value"
            >
              €
              0
            </b>
            <span
              className="property__price-text"
            >
               night
            </span>
          </div>
          <div
            className="property__inside"
          >
            <h2
              className="property__inside-title"
            >
              What's inside
            </h2>
            <Goods
              list={
                Array [
                  "",
                ]
              }
            />
          </div>
          <div
            className="property__host"
          >
            <h2
              className="property__host-title"
            >
              Meet the host
            </h2>
            <div
              className="property__host-user user"
            >
              <div
                className="property__avatar-wrapper property__avatar-wrapper--pro user__avatar-wrapper"
              >
                <img
                  alt=""
                  className="property__avatar user__avatar"
                  height="74"
                  src=""
                  width="74"
                />
              </div>
              <span
                className="property__user-name"
              />
            </div>
            <div
              className="property__description"
            >
              <p
                className="property__text"
              />
            </div>
          </div>
          <section
            className="property__reviews reviews"
          >
            <ConnectFunction
              id="0"
            />
            <form
              action="#"
              className="reviews__form form"
              method="post"
            >
              <label
                className="reviews__label form__label"
                htmlFor="review"
              >
                Your review
              </label>
              <div
                className="reviews__rating-form form__rating"
              >
                <input
                  className="form__rating-input visually-hidden"
                  id="5-stars"
                  name="rating"
                  type="radio"
                  value="5"
                />
                <label
                  className="reviews__rating-label form__rating-label"
                  htmlFor="5-stars"
                  title="perfect"
                >
                  <svg
                    className="form__star-image"
                    height="33"
                    width="37"
                  >
                    <use
                      xlinkHref="#icon-star"
                    />
                  </svg>
                </label>
                <input
                  className="form__rating-input visually-hidden"
                  id="4-stars"
                  name="rating"
                  type="radio"
                  value="4"
                />
                <label
                  className="reviews__rating-label form__rating-label"
                  htmlFor="4-stars"
                  title="good"
                >
                  <svg
                    className="form__star-image"
                    height="33"
                    width="37"
                  >
                    <use
                      xlinkHref="#icon-star"
                    />
                  </svg>
                </label>
                <input
                  className="form__rating-input visually-hidden"
                  id="3-stars"
                  name="rating"
                  type="radio"
                  value="3"
                />
                <label
                  className="reviews__rating-label form__rating-label"
                  htmlFor="3-stars"
                  title="not bad"
                >
                  <svg
                    className="form__star-image"
                    height="33"
                    width="37"
                  >
                    <use
                      xlinkHref="#icon-star"
                    />
                  </svg>
                </label>
                <input
                  className="form__rating-input visually-hidden"
                  id="2-stars"
                  name="rating"
                  type="radio"
                  value="2"
                />
                <label
                  className="reviews__rating-label form__rating-label"
                  htmlFor="2-stars"
                  title="badly"
                >
                  <svg
                    className="form__star-image"
                    height="33"
                    width="37"
                  >
                    <use
                      xlinkHref="#icon-star"
                    />
                  </svg>
                </label>
                <input
                  className="form__rating-input visually-hidden"
                  id="1-star"
                  name="rating"
                  type="radio"
                  value="1"
                />
                <label
                  className="reviews__rating-label form__rating-label"
                  htmlFor="1-star"
                  title="terribly"
                >
                  <svg
                    className="form__star-image"
                    height="33"
                    width="37"
                  >
                    <use
                      xlinkHref="#icon-star"
                    />
                  </svg>
                </label>
              </div>
              <textarea
                className="reviews__textarea form__textarea"
                id="review"
                name="review"
                placeholder="Tell how was your stay, what you like and what can be improved"
              />
              <div
                className="reviews__button-wrapper"
              >
                <p
                  className="reviews__help"
                >
                  To submit review please make sure to set 
                  <span
                    className="reviews__star"
                  >
                    rating
                  </span>
                   and describe your stay with at least 
                  <b
                    className="reviews__text-amount"
                  >
                    50 characters
                  </b>
                  .
                </p>
                <button
                  className="reviews__submit form__submit button"
                  disabled=""
                  type="submit"
                >
                  Submit
                </button>
              </div>
            </form>
          </section>
        </div>
      </div>
      <section
        className="property__map map"
      >
        <Map
          coordinates={Array []}
          currentCity=""
          selectedOffer={
            Array [
              "",
              "",
            ]
          }
        />
      </section>
    </section>
    <div
      className="container"
    >
      <section
        className="near-places places"
      >
        <h2
          className="near-places__title"
        >
          Other places in the neighbourhood
        </h2>
        <ConnectFunction
          classModCard="near-places__card"
          classModOffers={
            Array [
              "near-places__list",
            ]
          }
          offers={Array []}
        />
      </section>
    </div>
  </main>
</Layout>
`;
